src/internal/operations/manifests.go

```diff
@@ -0,0 +1,210 @@
+package operations
+
+import (
+	"context"
+
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/kopia/kopia/repo/manifest"
+	"github.com/pkg/errors"
+
+	"github.com/alcionai/corso/src/internal/connector/graph"
+	"github.com/alcionai/corso/src/internal/data"
+	"github.com/alcionai/corso/src/internal/kopia"
+	"github.com/alcionai/corso/src/internal/model"
+	"github.com/alcionai/corso/src/pkg/backup"
+	"github.com/alcionai/corso/src/pkg/logger"
+	"github.com/alcionai/corso/src/pkg/path"
+)
+
+type manifestFetcher interface {
+	FetchPrevSnapshotManifests(
+		ctx context.Context,
+		reasons []kopia.Reason,
+		tags map[string]string,
+	) ([]*kopia.ManifestEntry, error)
+}
+
+type manifestRestorer interface {
+	manifestFetcher
+	restorer
+}
+
+type getDetailsIDer interface {
+	GetDetailsIDFromBackupID(
+		ctx context.Context,
+		backupID model.StableID,
+	) (string, *backup.Backup, error)
+}
+
+// calls kopia to retrieve prior backup manifests, metadata collections to supply backup heuristics.
+func produceManifestsAndMetadata(
+	ctx context.Context,
+	mr manifestRestorer,
+	gdi getDetailsIDer,
+	reasons []kopia.Reason,
+	tenantID string,
+	getMetadata bool,
+) ([]*kopia.ManifestEntry, []data.Collection, bool, error) {
+	var (
+		metadataFiles = graph.AllMetadataFileNames()
+		collections   []data.Collection
+	)
+
+	ms, err := mr.FetchPrevSnapshotManifests(
+		ctx,
+		reasons,
+		map[string]string{kopia.TagBackupCategory: ""})
+	if err != nil {
+		return nil, nil, false, err
+	}
+
+	if !getMetadata {
+		return ms, nil, false, nil
+	}
+
+	// We only need to check that we have 1:1 reason:base if we're doing an
+	// incremental with associated metadata. This ensures that we're only sourcing
+	// data from a single Point-In-Time (base) for each incremental backup.
+	//
+	// TODO(ashmrtn): This may need updating if we start sourcing item backup
+	// details from previous snapshots when using kopia-assisted incrementals.
+	if err := verifyDistinctBases(ms); err != nil {
+		logger.Ctx(ctx).Warnw(
+			"base snapshot collision, falling back to full backup",
+			"error",
+			err,
+		)
+
+		return ms, nil, false, nil
+	}
+
+	for _, man := range ms {
+		if len(man.IncompleteReason) > 0 {
+			continue
+		}
+
+		bID, ok := man.GetTag(kopia.TagBackupID)
+		if !ok {
+			return nil, nil, false, errors.New("snapshot manifest missing backup ID")
+		}
+
+		dID, _, err := gdi.GetDetailsIDFromBackupID(ctx, model.StableID(bID))
+		if err != nil {
+			// if no backup exists for any of the complete manifests, we want
+			// to fall back to a complete backup.
```

Comment:
nit: this comment doesn't quite match the code (yes, I know this was just relocating code). The if-block says that if _any_ previous backup is missing the model/details we fallback to a full backup. If we want to make it such that _all_ previous backups need to be missing the model/details we'll need to check at the end of the loop. That's an optimization we can make in a different PR if we want

Reply:
I think you're saying the comment is correct and the behavior is wrong.  Both the comment and the code fallback on any missing backup, not if all backups are missing.

Either way, yes, punting.